add http status of Move SE

h07d0q opened the pull request:
I wanted to have a feedback for the standby state of my PTZOptics Move SE. Since I don't know a lot about VISCA and inquiries (and the issues read like it's impossible), I analyzed the network traffic of the webinterface. Using this information I saw that got is already implemented, but it doesn't support digest.

Please give feedback to my chosen solution and if other cameras are supported or need more variables. I can imagine the the variables can be created after the first call, so that cameras with other data response can be easily supported, but I don't know if it is needed.

jswalden commented:
Quick note, this is on my radar to look at carefully sometime in the next few days.  :slightly_smiling_face:  Tho, based on the workflow run result, it looks like you need to remove a trailing comma from `package.json` in the overall changes here.

jswalden commented:
FWIW you probably should just look at the [developer portal](https://ptzoptics.com/developer-portal/) as a first step, rather than reverse-engineering any kind of network protocol -- this interface is somewhat documented there (tho at a skim I don't see where this describes the JSON format): https://ptzoptics.imagerelay.com/share/PTZOptics-Command-List-HTTP-CGI-G3

h07d0q commented:
nvm the info about got - I was mixing things up. Just checked: the module uses TCPhelper for the VISCA commands.

I corrected the package.json - hoping the workflow will succeed.

h07d0q commented:
@jswalden please take a look at the las added features. It behaves like if you control the camera via webinterface. The webinterface polls the camera status every 3s or so.

following issues are affected:

#74

#45

#21

#2

h07d0q commented:
I have this version running on our old thin client since yesterday and the Linux system got unresponsive twice. We have connected two cameras.
Just now I had briefly access (to reboot it) and saw 100% CPU usage for this module. This isn't the case when running the connection for an hour or so. I set it very optimistic to 3s poll time for both cameras...
Maybe someone has an idea what could cause this?
I'll have a look at the logs - I remember I have seen errors when the camera was unresponsive or not reachable (mostly at night).

jswalden commented on src/instance.ts:
```diff
@@ -166,6 +232,159 @@ export class PtzOpticsInstance extends InstanceBase<PtzOpticsConfig> {
+				// check firmware version
+				try {
+					this.#digestClient2 = new DigestClient('', '')
+					const result = (await this.#digestClient2.fetch(`https://firmware.ptzoptics.com/${model}/RVU.json`, {
```
If PTZOptics actually documents this somewhere, I'm more than happy relying on it.  (But with a link here to the documentation.)

But if we're just reverse-engineering "APIs", there's a decent chance I'm going to put a foot down and say no.  I would rather not have the module working, then suddenly PTZOptics changes something and breaks existing module uses.

How much of all these things can be done just using documented APIs?  I see you can query "device info", "serial number", and other stuff using the APIs on the final page of https://ptzoptics.imagerelay.com/share/PTZOptics-Command-List-HTTP-CGI-G3 for example.  We should use those APIs, not discovered stuff.

h07d0q replied:
You are right, this is not documented anywhere and I understand if you say no.
I wrote it in that way if it doesn't work, the module would work anyway.
At least the user should have an option to opt out to query the ptz webservers, but I didn't implement this.
I simply checked the list on their website https://ptzoptics.com/latest-firmware-files/
Working queries:
| Device | DeviceID | URL |
|--------|--------|--------|
| PT12X-4K-xx-G3                   | F60.HI   | https://firmware.ptzoptics.com/F60.HI/RVU.json      |
| PT20X-4K-xx-G3                   | F51.HI   | https://firmware.ptzoptics.com/F51.HI/RVU.json      |
| PT30X-4K-xx-G3                   | F53.HI   | https://firmware.ptzoptics.com/F53.HI/RVU.json      |
| PT12X-SE-xx-G3                   | F59.HI   | https://firmware.ptzoptics.com/F59.HI/RVU.json      |
| PT20X-SE-xx-G3                   | F63.HI   | https://firmware.ptzoptics.com/F63.HI/RVU.json      |
| PT30X-SE-xx-G3                   | F64.HI   | https://firmware.ptzoptics.com/F64.HI/RVU.json      |
| PT12X-LINK-4K-xx                 | F60.HID  | https://firmware.ptzoptics.com/F60.HID/RVU.json     |
| PT20X-LINK-4K-xx                 | F51.HID  | https://firmware.ptzoptics.com/F51.HID/RVU.json     |
| PT30X-LINK-4K-xx                 | F53.HID  | https://firmware.ptzoptics.com/F53.HID/RVU.json     |
| HC20X-SIMPLTRACK3                | ST3      | https://firmware.ptzoptics.com/ST3/RVU.json         |
| PT-STUDIOPRO                     | P7.HI    | https://firmware.ptzoptics.com/P7.HI/RVU.json       |
| PT12X-STUDIO-4K-xx-G3           | P11.HI   | https://firmware.ptzoptics.com/P11.HI/RVU.json      |
| PT20X-STUDIO-4K-xx-G3           | P13.HI   | https://firmware.ptzoptics.com/P13.HI/RVU.json      |
| VL Fixed Camera/NDI Fixed Camera| O9.HI    | https://firmware.ptzoptics.com/O9.HI/RVU.json       |
| PT-JOY-G4                        | MP4      | https://firmware.ptzoptics.com/MP4/RVU.json         |
| PT-SUPERJOY-G1                   | MSJ      | https://firmware.ptzoptics.com/MSJ/RVU.json         |

not working:
| Device                                  | DeviceID | URL                                                                                      |
|-----------------------------------------|----------|-------------------------------------------------------------------------------------------|
| PT12X-SDI/NDI-xx-G2                     | F5.HI    | https://ptzoptics.nyc3.cdn.digitaloceanspaces.com/firmware/F5.HI/PT12X-SDI.NDI-xx-G2%20Firmware.zip |
| PT12X-SDI-xx-G2 (Non-PoE)               | F4.HI    | https://ptzoptics.com/wp-content/uploads/2023/11/PT12X-SDI-xx-G2-Non-PoE-Firmware.zip    |
| PT20X-SDI/NDI-xx-G2                     | F2.HI    | https://ptzoptics.nyc3.cdn.digitaloceanspaces.com/firmware/F2.HI/PT20X-SDI.NDI-xx-G2%20Firmware.zip |
| PT20X-SDI-xx-G2 (Non-PoE)               | F1.HI    | https://ptzoptics.com/wp-content/uploads/2023/11/PT20X-SDI-xx-G2-Non-PoE-Firmware.zip    |
| PT30X-SDI/NDI-xx-G2                     | F16.HI   | https://ptzoptics.nyc3.cdn.digitaloceanspaces.com/firmware/F16.HI/PT30X-SDI.NDI-xx-G2%20Firmware.zip |
| PT12X-USB-xx-G2                         | P8.HI    | https://ptzoptics.nyc3.cdn.digitaloceanspaces.com/firmware/P8.HI/PT12X-USB-xx-G2%20Firmware.zip |
| PT20X-USB-xx-G2                         | F14.HI   | https://ptzoptics.nyc3.cdn.digitaloceanspaces.com/firmware/F14.HI/PT20X-USB-xx-G2%20Firmware.zip |
| 12X Fixed Camera/NDI Fixed Camera       | P4.HI    | https://ptzoptics.nyc3.cdn.digitaloceanspaces.com/firmware/P4.HI/PT12X-(NDI)-ZCAM%20Firmware.zip |
| 20X Fixed Camera/NDI Fixed Camera       | P3.HI    | https://ptzoptics.nyc3.cdn.digitaloceanspaces.com/firmware/P3.HI/PT20X-(NDI)-ZCAM%20Firmware.zip |
| EPTZ Fixed Camera/NDI Fixed Camera      | 014.HI   | https://ptzoptics.nyc3.cdn.digitaloceanspaces.com/firmware/014.HI/PTEPTZ-(NDI)-ZCAM-G2%20Firmware.zip |

This list can contain not relevant models for this module.

As I only have one model and the result of `result.data.device_model` needs to be trimmed and split `devVersion = devVersion.split(' v')[1] || ''` I don't know if this works for all models but this can be resolved if we know how other cameras respond.

jswalden commented on src/instance.ts:
```diff
@@ -166,6 +232,159 @@ export class PtzOpticsInstance extends InstanceBase<PtzOpticsConfig> {
+							if (result?.data) {
+								if (firstRun) {
+									for (const key of Object.keys(result.data)) {
+										const varId = `HTTP${key}`
```
Oh, hmm, you're directly tying the variable IDs to what's in this CGI API.  I think I would prefer you don't do that, and that you specifically select out the keys you're using and manually map them to the variable IDs.  Implicit dependencies (especially coming from an API that just returns more or less JSON) make code harder to understand.

jswalden commented on src/instance.ts:
```diff
@@ -166,6 +232,159 @@ export class PtzOpticsInstance extends InstanceBase<PtzOpticsConfig> {
+					void (async () => {
+						if (isRequestInProgress) return // skip if already/still running
+						// Get tally status
+						try {
```
The two halves of this should be broken into their own sub-functions, too.

jswalden commented on yarn.lock:
```diff
@@ -1238,6 +1238,13 @@ __metadata:
   languageName: node
   linkType: hard
 
```
Please make sure to do a final `yarn install` whose results are included in the next updates -- `yarn.lock` changes for me when I check out your branch and run `yarn install` in it.

Also, I'd appreciate you rebasing this on top of `master`, because I'll eventually want that to have happened, and if it's not done already, there's minor risk something won't work when it's attempted in the Github landing process.

jswalden reviewed (changes requested):
Sorry for the delay on this, my immediate priority has been getting the `allenheath-sq` module in shape to release (for the first time in years), and then after that trying to wrap up a couple changes in `youtube-live` that been in demand longer than this.

I'm hopeful if you can deal with my feedback, we can get this done this week and get a release out to get this in Companion 4.0.  But if not, given that 4.0 introduces some decoupling of Companion version from Companion modules and allows downloading newer releases, I think it'll be less of an irritation now than it used to be.

h07d0q commented on src/instance.ts:
```diff
@@ -1,4 +1,5 @@
 import { InstanceBase, InstanceStatus, type SomeCompanionConfigField } from '@companion-module/base'
+import DigestClient from 'digest-fetch'
```
I looked for newer, maintained alternatives and only found [undici](https://github.com/nodejs/undici) which supports interceptors. I tried it with [mikaelkaron/undici-digest-interceptor](https://github.com/mikaelkaron/undici-digest-interceptor) but [didn't get it to work](https://github.com/bitfocus/companion-module-ptzoptics-visca/commit/ab334f4e1d12fe5291c1fe9c3d1bd51d200a86b8) after multiple evenings 😞
That's why I went back to [digest-fetch](https://github.com/devfans/digest-fetch) (wrapper/plugin for [node-fetch](https://github.com/node-fetch/node-fetch) latest release 2023-07-25)

h07d0q commented:
Hi @jswalden,

Chris from #73 and me managed to get a working version of his G2  model and my G3 Move SE using the reboot action.
The G2 model uses MD5 instead SHA256 digest authorization and shares the URL from the documentation:
`http://[camera ip]/cgi-bin/param.cgi?get_device_conf`
with the same response:
`devname="HD Camera" versioninfo="X9.02.29" device_model=" V.F2" serial="vf292------"`
which should be a PT20X-SDI/NDI-xx according to this [list](https://github.com/bitfocus/companion-module-ptzoptics-visca/pull/73#discussion_r2130291828)
`devname="HD Camera" versioninfo="X8.02.29" device_model=" V.F3" serial="vf292------"`
Should be the same but 30x zoom like Chris is saying: PT30X-SDI/NDI-xx

Regarding the reboot action:
Chris uses v3.6.3 since some days https://github.com/bitfocus/companion-module-ptzoptics-visca/pull/73/commits/b32c691adab4f1b7ade4b3dc966ed035fa181906
It calls the same URL like the web interface, and the G3 models.

If possible, I wouldn't implement a model selector and leave it to the user (adding static info which models are supported).

In https://github.com/bitfocus/companion-module-ptzoptics-visca/issues/77#issuecomment-3251015456 I added the tally action, regardless if there is an old documented VISCA command, as this was straight forward to implement. There I added an info to the action, which cameras support this command.

Sorry that this PR ist that massive and I don't see how I should narrow this down to this action, since all requires the HTTP commands... 
summarized function:
- HTTP commands using digest authentication and authorization detection should work for G2 and G3
- Reboot action
- Tally mode action (two G3 models)
- Variables when polling (don't think this works for G2)

I already changed some parts of your requested changes - please re-review again :)

ChrisSW58 commented:
@h07d0q  and @jswalden - Merry Christmas/Happy Holidays to you both, and I hope you had a wonderful time. May I ask, is there any way I could get the Zoom speed controls implemented on this module? Although they are stated in the developer portal, they have not been implemented (although the P/T Speed has been implemented).The feature is available and works via the direct camera URL screen. If no plans to implement it, could you please advise what the fields in the HTTP custom command action would contain? Many thanks.

ChrisSW58 commented:
> Hi Chris!
> 
> I don't remember I implemented a custom http command action... Where did you see the commands in the developer portal?
> 
> For my camera the URL is a POST request to http://ipaddress/api/v1/speeds which responds: `{"focus":5, "motionSync":"on", "pan":10, "tilt":10, "zoom":6}` One also can set speeds with the same payload and responses `{msg: "operate success"}`
> 
> Can you confirm this using the method I showed you?
> 
> BTW: I found an official example for sending commands: https://github.com/PTZOptics/Code-Examples/blob/main/HTTP%20API/node/example-http.js The more I look at my code, the more I think it would be better to outsource the HTTP API to an own module...

H Eugen - if you look at the PTZ Command List at the start of this post, you will see that on the Zoom In and Zoom out commands, the last variable (optional I assume) is Zoom speed. (I\ve copied the comment in the attachment).  In the direct URL browser window for the camera, I have a dialog box where I can change the zoom speed, so I 'assume' if I could pass that variable to the camera on the Zoom In and Zoom Out commands, then I may be able to control the zoom speed.

If a camera is working successfully without the parameter, then passing it across should not affect it...correct? In which case, being able to make an action call to set /increase/decrease the Zoom speed, which would then be passed on the Zoom calls, should not affect this cameras which don't support that capability... Yes?

<img width="1154" height="265" alt="Screenshot 2025-12-28 at 1 47 32 am" src="https://github.com/user-attachments/assets/61889689-8cae-4dbe-86d8-a555cd1977c8" />

ChrisSW58 commented:
> Yes this should work but that means it will also change the actual zoom, focus or PT, not only set the speed like when changing the speed in the web interface.

Yes - but if it's a variable you simply add to the camera call, then it's an additional optional field in the Zoom In and Zoom Out commands.

> As I wouldn't like to implement PTZ functions via HTTP API - it would be better to implement this via the VISCA IP commands since the command is supported there too. Please open a new issue as feature request.

Done - many thanks!
